### ISSUE TITLE
Work around broken OSGi implementations Bundle.getEntry() behavior returning with unescaped URLs

### ISSUE THREAD

**Issue by jbrazeau** [eclipse/jetty.project#2575]
For example, if you package an eclipse application that embeds jetty, and if you rely on `org.eclipse.jetty.osgi.boot` to start Jetty, if your eclipse installation path contains a space for example, Jetty default bootstrap fails with the following error :
```
!STACK 0
java.lang.IllegalStateException: No such jetty server config file: file:/Users/jbrazeau/a path with a space/MyEclipseApp.app/Contents/Eclipse/configuration/org.eclipse.osgi/215/0/.cp/jettyhome/etc/jetty.xml; BadResource=java.net.URISyntaxException: Illegal character in path at index 22: file:/Users/jbrazeau/a path with a space/MyEclipseApp.app/Contents/Eclipse/configuration/org.eclipse.osgi/215/0/.cp/jettyhome/etc/jetty.xml
	at org.eclipse.jetty.osgi.boot.internal.serverfactory.ServerInstanceWrapper.configure(ServerInstanceWrapper.java:153)
	at org.eclipse.jetty.osgi.boot.internal.serverfactory.DefaultJettyAtJettyHomeHelper.startJettyAtJettyHome(DefaultJettyAtJettyHomeHelper.java:179)
	at org.eclipse.jetty.osgi.boot.JettyBootstrapActivator.start(JettyBootstrapActivator.java:82)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:779)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:772)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:729)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:933)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:309)
	at org.eclipse.osgi.container.Module.doStart(Module.java:581)
	at org.eclipse.osgi.container.Module.start(Module.java:449)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1634)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.incStartLevel(ModuleContainer.java:1614)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.doContainerStartLevel(ModuleContainer.java:1585)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1528)
	at org.eclipse.osgi.container.ModuleContainer$ContainerStartLevel.dispatchEvent(ModuleContainer.java:1)
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:230)
	at org.eclipse.osgi.framework.eventmgr.EventManager$EventThread.run(EventManager.java:340)

```
Jetty retrieves dynamically the `org.eclipse.jetty.osgi.boot` bundle URL in order to load the `jetty.xml` configuration file that resides in this bundle. Then the URL is converted to a `java.net.URI` in the `PathResource` class as you can see here : https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java#L285

This works perfectly unless the `org.eclipse.jetty.osgi.boot*.jar` full path contains a space (or more precisely any character that is considered as invalid in a URI like `{`, `[`, ...).

That situation leads to a `BadResource` instance here (instead of a valid `PathResource`) : https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java#L123

And then, jetty complains about an invalid URL and stops.

It is true that a path that contains a `[` is strange, but a path that contains a space is classic (especially, but not only, for windows users).

Personnaly, I would expect jetty to be able to start even in a folder that contains a space (I feel that at the end, this looks like a strange limitation for the end user).

**Comment by jbrazeau** [comment]
Note #1 : I met this issue with jetty `9.4.7` (`9.4.7.v20170914` more precisely) but as far as I can see the 9.4.x branch head still have the same problem (see mentionned sources links above)
Not #2 : I feel that 10.0.x has the same issue if I take a look at : 
* https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java#L285
* https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java#L123

**Comment by jbrazeau** [comment]
As far as I can see, a very simple fix may solve this issue.

`PathResource` class is dedicated to file paths, so the following `URL` based constructor :
```java
public PathResource(URL url) throws IOException, URISyntaxException
{
    this(url.toURI());
}
```
may simply be replaced by :
```java
public PathResource(URL url) throws IOException, URISyntaxException {
	this(toURI(url));
}

private static URI toURI(URL fileUrl) {
	if (!fileUrl.getProtocol().equalsIgnoreCase("file")) {
		throw new IllegalArgumentException("not file: scheme");
	} else {
		return new File(fileUrl.getFile()).toURI();
	}
}
```
Adding the `file` protocol check is not a problem as this is the same check as in the `URI` based constructor (see this line of code https://github.com/eclipse/jetty.project/blob/jetty-10.0.x/jetty-util/src/main/java/org/eclipse/jetty/util/resource/PathResource.java#L235). This check is done to be sure that it is possible to use safely `URL#getFile` in the following lines of code.

And then using `File#toURI` algorithm to turn the path to a `URI` (instead of the `URL#toURI` algorithm) is far more secure & robust (space character does lead to a failure anymore, it is correctly handled).

**Comment by olamy** [comment]
@jbrazeau can you make a pull request? so we can verify with all the tests?

**Comment by jbrazeau** [comment]
Here is the pull request : https://github.com/eclipse/jetty.project/pull/2587

Note #1 : instead of reusing the `URI` based constructor, I reused the `File` based constructor (which is better I think, and at least equivalent).
Note #2 : Instead of duplicating the scheme/protocol check (and the associated error message), I reused this little portion of code through the `checkFileURIScheme` method.

**Comment by joakime** [comment]
I'm adding some tests of PathResource and URL's that have directories with spaces to verify that PathResource behavior.

**Comment by joakime** [comment]
PR #2590 tests the theory that PathResource(URL) is at fault.

On Windows 10, these tests pass.
CI will tell us about Linux.

**Comment by jbrazeau** [comment]
Excuse me but I think that my explanations are not clear enough. I'm really sorry for that.

I will try to explain it differently.

First, please notice that I'm not saying that URL object doesn't handle encoded characters (such as spaces) consistently across OS's. I'm not complaining about any OS dependant problem. 

As far as I can see, the test case that you have created does not cover the issue I'm trying to describe because the URL that you use is derivated from a URI (through `URI#toUrl` as we can see here : https://github.com/eclipse/jetty.project/commit/d290e74b068e32007dd8b19216b9cb97b05f0103#diff-acc2d67f85bbbd3a5e8dba774312834aR264)

Now consider this simple code :
```java
// Create a sample file with a name that contains spaces
String path = "/tmp/my file with spaces.txt";
File file = new File(path);
FileOutputStream fout = new FileOutputStream(file);
fout.write("Hello world\n".getBytes());
fout.close();
System.out.println(path + " exists ? " + file.exists());

// Try to create a new path resource from the File object : everything
// works fine
new PathResource(file);

// Try to create a new path resource from the File object toURI :
// everything works fine
new PathResource(file.toURI());

// Try to create a new path resource from the File object
// toURI().toURL() : everything works fine
// (when the URL is derivated from a URI the spaces
// are "URL encoded" (with %20)
new PathResource(file.toURI().toURL());

// Now let's consider a URL that targets the same path, but that is
// built manually
URL url = new URL("file:" + path); // That URL contains spaces, no %20
// This path exist, the URL is valid as we can open it and read
// the file content
System.out.println("File content : ");
try (InputStream in = url.openConnection().getInputStream()) {
	IO.copy(in, System.out); // outputs "Hello world" in the console
}
// Unfortunately, even if the URL can be read, it is not possible
// to create a new PathResource from this URL like this :
new PathResource(url); // See exception below
/*
java.net.URISyntaxException: Illegal character in path at index 12: file:/tmp/my file with spaces.txt
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.checkChars(URI.java:3021)
	at java.net.URI$Parser.parseHierarchical(URI.java:3105)
	at java.net.URI$Parser.parse(URI.java:3053)
	at java.net.URI.<init>(URI.java:588)
	at java.net.URL.toURI(URL.java:946)
	at org.eclipse.jetty.util.resource.PathResource.<init>(PathResource.java:285)
*/
```
First I create a simple file in `/tmp` folder, then I create a `PathResource`through different scenarios : if we use a file, a URI or even a URI that has been derivated from a URI, everything goes well.

The problem appears with URLs that have been built differently : in the previous code, I create the last URL manually, and to show you that the URL is valid, I read the URL to read the underlying file wich works perfectly.

Unfortunately, this URL cannot be converted to a URI. It fails with a `java.net.URISyntaxException` due to the fact that this URL contains a space that is considered as an invalid URI character. But the targeted file really exists !

Now please consider that Eclipse URL management (`platform://`, `bundleentry://`, ...) typically creates URL that are similar to the URL that has been built manually in the code above. It typically can contain spaces (not "URL encoded") if the corresponding path contains spaces.

In `org.eclipse.jetty.osgi.boot.utils.internal.DefaultFileLocatorHelper` jetty uses reflection to call `org.eclipse.osgi.storage.url.BundleURLConnection#getFileURL` -> This can typically result in URLs that may contain spaces.

That is why, according to me, turning the URL to a File, and then turning the File to a URI is better than turning a URL to a URI because saying that a URL cannot be converted to a URI doesn't mean that the targeted file doesn't exist and it doesn't mean that the URL is invalid.

**Comment by joakime** [comment]
``` java
URL url = new URL("file:" + path); // That URL contains spaces, no %20
```

_(Unencoded spaces are just the tip of the iceberg)_

That is an invalid URL and will throw exception on some JVM's and even on some OS's. (not to mention the other fault with path separators on some OSs as well).

If this is the case for `Bundle.getEntry()`, where it returns a bad URL, that isn't encoded, then the fix must exist solely on the jetty-osgi side, not within `PathResource` as that's used in far too many situations outside of the broken osgi layer.
But even then, there's no real hope for Jetty to "fix" this bad url encoding.
The URL returned from `Bundle.getEntry()` **must** be encoded properly to have any hope of this working.  

In short, the raw URL format, without encoding, is invalid and will not work reliably.

Here's a testcase for you to ponder.

``` java
package uri;

import static java.nio.charset.StandardCharsets.UTF_8;
import static org.junit.Assert.assertNotNull;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Writer;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;

import org.junit.BeforeClass;
import org.junit.Test;

public class UrlBadEncodingTest
{
    private static Path basePath;
    private static ArrayList<String> encodedUris = new ArrayList<>();
    private static ArrayList<String> rawUris = new ArrayList<>();

    @BeforeClass
    public static void setUp() throws IOException
    {
        Path tmpPath = Paths.get(System.getProperty("java.io.tmpdir"));
        basePath = tmpPath.resolve("demo");
        mkdir(basePath);
        Path appValue = basePath.resolve("app%2fvalue");
        mkdir(appValue);
        Path jettyXmlFile = appValue.resolve("jetty.xml");
        write(jettyXmlFile, "Hello App Value");

        encodedUris.add(jettyXmlFile.toUri().toASCIIString());
        rawUris.add("file:" + separators(jettyXmlFile.toString()));

        Path ciDir = basePath.resolve("ci");
        mkdir(ciDir);
        Path jobsDir = ciDir.resolve("jobs");
        mkdir(jobsDir);
        Path projectDir = jobsDir.resolve("project 20%%2fbuild jdk11");
        mkdir(projectDir);
        Path projectFile = projectDir.resolve("Jenkinsfile");
        write(projectFile, "Example Jenkinsfile");

        encodedUris.add(projectFile.toUri().toASCIIString());
        rawUris.add("file:" + separators(projectFile.toString()));

        /*  At this point you have a directory structure like this ...
         *
         * $ find demo/
         * demo/
         * demo/app%2fvalue/
         * demo/app%2fvalue/jetty.xml
         * demo/ci/
         * demo/ci/jobs/
         * demo/ci/jobs/project 20%%2fbuild jdk11/
         * demo/ci/jobs/project 20%%2fbuild jdk11/Jenkinsfile
         */
    }

    private static void mkdir(Path path) throws IOException
    {
        if(Files.exists(path))
            return;
        Files.createDirectory(path);
    }

    public static String separators(String path)
    {
        StringBuilder ret = new StringBuilder();
        for (char c : path.toCharArray())
        {
            if ((c == '/') || (c == '\\'))
            {
                ret.append('/'); // always use URL/URL path separators
            }
            else
            {
                ret.append(c);
            }
        }
        return ret.toString();
    }


    @Test
    public void testEncodedWithURI() throws URISyntaxException, IOException
    {
        for (String encodedUri : encodedUris)
        {
            System.out.println(encodedUri);
            URI uri = new URI(encodedUri);
            String contents = readContents(uri);
            assertNotNull(contents);
        }
    }

    @Test
    public void testEncodedWithURL() throws IOException
    {
        for (String encodedUri : encodedUris)
        {
            System.out.println(encodedUri);
            URL url = new URL(encodedUri);
            String contents = readContents(url);
            assertNotNull(contents);
        }
    }

    @Test
    public void testRawWithURI() throws URISyntaxException, IOException
    {
        for (String rawUri : rawUris)
        {
            System.out.println(rawUri);
            URI uri = new URI(rawUri);
            String contents = readContents(uri);
            assertNotNull(contents);
        }
    }

    @Test
    public void testRawWithURL() throws IOException
    {
        for (String rawUri : rawUris)
        {
            System.out.println(rawUri);
            URL url = new URL(rawUri);
            String contents = readContents(url);
            assertNotNull(contents);
        }
    }

    private String readContents(URI uri) throws IOException
    {
        return readContents(uri.toURL());
    }

    private String readContents(URL url) throws IOException
    {
        try (InputStream input = url.openStream();
             InputStreamReader reader = new InputStreamReader(input, UTF_8);
             BufferedReader buf = new BufferedReader(reader))
        {
            return buf.readLine();
        }
    }

    private static void write(Path file, String text) throws IOException
    {
        try (Writer out = Files.newBufferedWriter(file, UTF_8))
        {
            out.write(text);
        }
    }
}
```

**Comment by janbartel** [comment]
@joakime it's pretty clear that if you pass in a url that contains spaces to new PathResource(url), then a URISyntaxException will be thrown in the constructor. However, I don't think the fix should go in PathResource - if it is passed a bad URL I think it's entirely reasonable that it throws an exception.

IMHO the fix is simply in the osgi class ServerInstanceWrapper to ensure that we escape any spaces to %20 in the url before passing it to Resource.newResource(url).

**Comment by joakime** [comment]
@janbartel point is, you can't arbitrarily encode, this problem isn't limited to just spaces.

Example:

```
$ find demo/
demo/
demo/app%2fvalue/
demo/app%2fvalue/jetty.xml
demo/ci/
demo/ci/jobs/
demo/ci/jobs/project 20%%2fbuild jdk11/
demo/ci/jobs/project 20%%2fbuild jdk11/Jenkinsfile
```

Those 2 files will look like ...

```
file:/tmp/demo/app%2fvalue/jetty.xml
file:/tmp/demo/ci/jobs/project 20%%2fbuild jdk11/Jenkinsfile
```

It might make more sense to take the bad URL from `Bundle.getEntry()`, pass it to `new URLResource(url, false, useCaches)` and not let the discovery mechanism for "file", and "jar:file" do its thing.
However, even that isn't a great solution, as the two above URLs prove (those wont work).

Even taking the `URL.getPath()` to pass to `new File(String)` wont work when Eclipse fixes their `Bundle.getEntry()` behavior to start encoding, suddenly that technique will stop working too.

**Comment by jbrazeau** [comment]
> Even taking the URL.getPath() to pass to new File(String) wont work when Eclipse fixes their Bundle.getEntry() behavior to start encoding, suddenly that technique will stop working too.

wow, good shot : you're perfectly right...

**Comment by janbartel** [comment]
The point is, we cannot know whether or not the url returned to us from Eclipse has been properly escaped or not. The only thing we can safely do is to convert space to %20. That will cover 99% of cases. I think we should do that.

**Comment by joakime** [comment]
Still safer to just use the `URLResource` directly, then we don't care what `Bundle.getEntry()` does now or in the future.

**Comment by joakime** [comment]
@jbrazeau can you build/use Jetty from branch `jetty-9.4.x-issue-2575-pathresource-urls-with-spaces` and see if the issue still persists for you?

**Comment by jbrazeau** [comment]
@joakime : if I can, I will do. But I am not sure.

The `org.eclipse.jetty.util_9.4.7.v20170914.jar` bundle is signed, so that I cannot replace the class file for a "quick" test (assuming I manage to compile the only class that has been modified by the patch).

So as far as I can see I would have to build a full Jetty P2 repository (like the official one here : http://download.eclipse.org/jetty/updates/jetty-bundles-9.x/). Can you tell me how you build your P2 repository ? Is the default maven build (using the root pom of jetty) will do the job ?

**Comment by jbrazeau** [comment]
Or maybe I can use hot code replacement in remote debug. I'll give it a try and give you a feedback as soon as possible.

**Comment by joakime** [comment]
@jbrazeau hold off on this for a bit.  A few things discovered that still need addressing.

**Comment by joakime** [comment]
Looks like `ServerInstanceWrapper#extractFiles()` has the same fundamental issue/flaw.

https://github.com/eclipse/jetty.project/blob/b75a2df67935aec18b5896aca70502adf4a0a267/jetty-osgi/jetty-osgi-boot/src/main/java/org/eclipse/jetty/osgi/boot/internal/serverfactory/ServerInstanceWrapper.java#L454-
L459

**Comment by joakime** [comment]
@janbartel writing a unit test to replicate this isn't really possible with paxexam and maven in the way.

However, you can manually setup the conditions so that the jetty-osgi/test-jetty-osgi/ will fail.

You have to do 2 things.

1. Setup a local maven repository directory location with spaces.
2. Git Clone the eclipse/jetty.project repository into a directory with spaces.

``` xml
<?xml version="1.0" ?>

<settings xmlns="http://maven.apache.org/POM/4.0.0"
  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
  xsi:schemaLocation="http://maven.apache.org/POM/4.0.0
                      http://maven.apache.org/xsd/settings-1.0.0.xsd">
  <localRepository>C:/temp/repo with spaces/</localRepository>
</settings>
```

```
$ git clone git@github.com:eclipse/jetty.project.git "jetty with spaces"
```

If paxexam uses the local repository, then the `Bundle.getEntry()` will return the new maven local repository URLs with spaces.
If paxexam uses the maven reactor, then the cloned directory with spaces kicks in for URLs with spaces.

**Comment by joakime** [comment]
paxexam and osgi *really* don't like it when spaces are present.

here's a command line build of just `jetty-osgi/test-jetty-osgi` with `$ mvn clean install` when the maven local repository only has been changed to use spaces.

```
[INFO] Running org.eclipse.jetty.osgi.test.TestJettyOSGiBootContextAsService
2018-05-30 13:51:47,597 [ INFO][org.ops4j.pax.exam.spi.DefaultExamSystem] Pax Exam System (Version: 4.11.0) created.
2018-05-30 13:51:47,600 [DEBUG][org.ops4j.store.intern.TemporaryStore] Storage Area is C:\code\jetty\jetty.project-9.4.x\jetty-osgi\test-jetty-osgi\target\1527706307598-0
2018-05-30 13:51:47,644 [ INFO][org.ops4j.pax.exam.junit.impl.ProbeRunner] creating PaxExam runner for class org.eclipse.jetty.osgi.test.TestJettyOSGiBootContextAsService
2018-05-30 13:51:47,650 [DEBUG][org.ops4j.pax.exam.spi.PaxExamRuntime] Found TestContainerFactory: org.ops4j.pax.exam.forked.ForkedTestContainerFactory
2018-05-30 13:51:47,659 [DEBUG][org.eclipse.jetty.util.log] Logging to org.slf4j.impl.Log4jLoggerAdapter(org.eclipse.jetty.util.log) via org.eclipse.jetty.util.log.Slf4jLog 2018-05-30 13:51:47,663 [ INFO][org.eclipse.jetty.util.log] Logging initialized @910ms to org.eclipse.jetty.util.log.Slf4jLog
2018-05-30 13:51:47,703 [DEBUG][org.ops4j.pax.exam.spi.DefaultExamSystem] creating default probe
2018-05-30 13:51:47,838 [DEBUG][org.ops4j.pax.exam.spi.reactors.ReactorManager] No User defined probe hook found
2018-05-30 13:51:47,839 [DEBUG][org.ops4j.pax.exam.spi.intern.DefaultTestAddress] NEW ADDRESS= PaxExam-812440bc-7f8c-4b5d-bc1b-12ca287ef4ff parent=null root=[TestAddress:PaxExam-812440bc-7f8c-4b5d-bc1b-12ca287ef4ff root:PaxExam-812440bc-7f8c-4b5d-bc1b-12ca287ef4ff] args=[Ljava.lang.Object;@22ef9844
2018-05-30 13:51:47,840 [DEBUG][org.ops4j.pax.exam.spi.DefaultExamReactor] Staging reactor with probes: 1 using strategy: org.ops4j.pax.exam.spi.reactors.PerMethod@611889f4 2018-05-30 13:51:47,840 [DEBUG][org.ops4j.store.intern.TemporaryStore] Storage Area is C:\code\jetty\jetty.project-9.4.x\jetty-osgi\test-jetty-osgi\target\1527706307598-0
2018-05-30 13:51:47,871 [DEBUG][org.ops4j.pax.exam.spi.intern.DefaultTestAddress] NEW ADDRESS= PaxExam-b20629e2-97a2-4907-b4cb-8d3abcd62bef parent=[TestAddress:PaxExam-812440bc-7f8c-4b5d-bc1b-12ca287ef4ff root:PaxExam-812440bc-7f8c-4b5d-bc1b-12ca287ef4ff] root=[TestAddress:PaxExam-812440bc-7f8c-4b5d-bc1b-12ca287ef4ff root:PaxExam-812440bc-7f8c-4b5d-bc1b-12ca287ef4ff] args=[Ljava.lang.Object;@729d991e
2018-05-30 13:51:47,874 [ INFO][org.ops4j.pax.exam.junit.impl.ProbeRunner] running test class org.eclipse.jetty.osgi.test.TestJettyOSGiBootContextAsService
2018-05-30 13:51:47,877 [DEBUG][org.ops4j.pax.exam.junit.impl.ProbeRunner] Invoke testContextHandlerAsOSGiService @ [TestAddress:PaxExam-b20629e2-97a2-4907-b4cb-8d3abcd62bef root:PaxExam-812440bc-7f8c-4b5d-bc1b-12ca287ef4ff] Arguments: [Ljava.lang.Object;@22ef9844
2018-05-30 13:51:47,877 [DEBUG][org.ops4j.store.intern.TemporaryStore] Storage Area is C:\code\jetty\jetty.project-9.4.x\jetty-osgi\test-jetty-osgi\target\1527706307598-0
2018-05-30 13:51:47,884 [DEBUG][org.ops4j.pax.exam.forked.ForkedFrameworkFactory] using RMI registry at port 21000
2018-05-30 13:51:47,931 [DEBUG][org.ops4j.exec.DefaultJavaRunner] Start command line [[C:\Program Files\Java\jdk-9.0.4/bin/java, -Dpax.swissbox.framework.rmi.port=21000, -Dorg.eclipse.jetty.LEVEL=WARN, -Djava.protocol.handler.pkgs=org.ops4j.pax.url, -Dorg.ops4j.pax.logging.DefaultServiceLog.level=WARN, -Dorg.ops4j.pax.url.mvn.localRepository=C:\temp\repo with spaces, -Dpax.swissbox.framework.rmi.name=ExamRemoteFramework-a32d8de9-6de4-443e-9fc7-b75afbde832a, -Djetty.http.port=0, -Dpax.exam.invoker=junit, -Dpax.exam.service.timeout=10000, -Djetty.base=C:\code\jetty\jetty.project-9.4.x\jetty-osgi\test-jetty-osgi\src\test\config, -Djava.rmi.server.hostname=127.0.0.1, -Djetty.etc.config.urls=file:/C:/code/jetty/jetty.project-9.4.x/jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty.xml;file:/C:/code/jetty/jetty.project-9.4.x/jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty-http-boot-context-as-service.xml;file:/C:/code/jetty/jetty.project-9.4.x/jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty-deployer.xml;file:/C:/code/jetty/jetty.project-9.4.x/jetty-osgi/test-jetty-osgi/src/test/config/etc/jetty-testrealm.xml, -Djetty.home=C:\code\jetty\jetty.project-9.4.x\jetty-osgi\test-jetty-osgi\src\test\config, -cp, /C:/temp/repo with spaces/org/eclipse/platform/org.eclipse.osgi/3.12.50/org.eclipse.osgi-3.12.50.jar;/C:/temp/repo with spaces/org/ops4j/pax/swissbox/pax-swissbox-framework/1.8.2/pax-swissbox-framework-1.8.2.jar;/C:/temp/repo with spaces/org/ops4j/pax/swissbox/pax-swissbox-tracker/1.8.2/pax-swissbox-tracker-1.8.2.jar, org.ops4j.pax.swissbox.framework.RemoteFrameworkImpl, -Forg.osgi.framework.bootdelegation=sun.*,org.xml.sax,org.xml.*,org.w3c.*,javax.xml.*, -Forg.osgi.framework.storage=C:\code\jetty\jetty.project-9.4.x\jetty-osgi\test-jetty-osgi\target\1527706307598-0]]
2018-05-30 13:51:47,931 [DEBUG][org.ops4j.exec.DefaultJavaRunner] Starting platform process.
2018-05-30 13:51:47,934 [DEBUG][org.ops4j.exec.DefaultJavaRunner] Wrapping stream I/O.
2018-05-30 13:51:47,936 [DEBUG][org.ops4j.exec.DefaultJavaRunner] Added shutdown hook.
2018-05-30 13:51:47,936 [ INFO][org.ops4j.exec.DefaultJavaRunner] DefaultJavaRunner completed successfully

Error: Could not find or load main class org.ops4j.pax.swissbox.framework.RemoteFrameworkImpl
Caused by: java.lang.ClassNotFoundException: org.ops4j.pax.swissbox.framework.RemoteFrameworkImpl
java.io.IOException: The pipe is being closed
        at java.base/java.io.FileOutputStream.writeBytes(Native Method)
        at java.base/java.io.FileOutputStream.write(FileOutputStream.java:332)
        at java.base/java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:81)
        at java.base/java.io.BufferedOutputStream.flush(BufferedOutputStream.java:142)
        at org.ops4j.io.Pipe$Pump.run(Pipe.java:157)
```

**Comment by jbrazeau** [comment]
For eclipse, I agree that there is definitely a problem.

Consider the following (strange) case : I have a plugin in which I have a file that has a (strange) name : `strangeNameWith%20.txt` (this name is said to be strange because it contains the `%20` as you can see).

Imagine I use eclipse API to retrieve a URL that targets this file, I do someting like this : 
`URL platformUrl = FileLocator.find(new URL("platform:/plugin/mytestplugin/strangeNameWith%20.txt")); `

Eclipse returns something like this : `bundleentry://25.fwk1072601481/strangeNameWith%20.txt`.
I can resolve it to get a more comprehensive URL with : 
`URL resolvedUrl = FileLocator.resolve(platformUrl);`

And I get : `file:/mypath/myworkspace/myplugin/strangeNameWith%20.txt` in which we recognize the full path of our file. But the problem is that the '%' has not been encoded. So if I turn this URL to an URI (`URL#toURI)`, the resulting URI does not target my real file (`%20` is replaced by a space : `file:/mypath/myworkspace/myplugin/strangeNameWith .txt`).

In that particular case, doing `new File(resolvedUrl.getFile()).toURI().toURL()` would help to get the good URL. It should work because this test shows that **eclipse does not encode at all bundle entry URLs**. 

That's why, I'm starting to think that the good fix (for eclipse) would probably consist in udpating the following methods of the `org.eclipse.jetty.osgi.boot.utils.internal.DefaultFileLocatorHelper` class : 
* `getFileURL(URL)`
* `getLocalURL(URL)`
* maybe `getBundleInstallLocation(Bundle)` ? (I'm not sure, it would have to be checked)

In both `getFileURL(URL)` & `getLocalURL(URL)`, I think it would be good to keep the resulting URL and perform a `new File(url.getFile()).toURI().toURL()` before returning the result.

To prevent from failing the day eclipse start to encode URLs (as @joakime rightly mentioned), it would be possible to have a static boolean field in the class that performs a simple eclipse based URL tretrieval to check if eclipse encodes its URLs (for example by putting in the OSGi bunde a resource with a space in the name and checking if the URL contains a space or `%20`).

The boolean field would condition the URL post processing in * `getFileURL(URL)` & `getLocalURL(URL)` (and maybe `getBundleInstallLocation(Bundle)` ? but this has to be checked).

I don't know PaxExam, so I don't know if what is described above can help, but if there is no link between Eclipse & paxexam, it will not for sure...

**Comment by joakime** [comment]
Merged into `jetty-9.4.x` and released with Jetty 9.4.11.v20180605
